### ISSUE TITLE
Consider: GitHub actions for CI

### ISSUE THREAD

**Issue by devjgm** [googleapis/google-cloud-cpp#3865]
**No rush or urgency here**

We should look into using GitHub actions as our CI runner. It looks like it supports most/all of what we need. It supports running tests on linux, windows, and macos. It supports using docker images. It supports caching data between runs. It supports nice UI integration, and real-time streaming of logs in the UI (awesome!).

Some obvious concerns are:
* I believe we can only cache 5GB per repo, which I imagine would not be enough.
* There may be other resource limits that we'd hit.

We can, of course, also consider running our own GitHub Action runners, but that's more work.

Just something that we might want to consider since Kokoro sometimes bites us.

**Comment by coryan** [comment]
@devjgm should we summarize your findings and close this?

**Comment by coryan** [comment]
Ping ?

**Comment by devjgm** [comment]
Summary:

* I added basic GitHub actions support in https://github.com/googleapis/google-cloud-cpp/pull/4333
* Configuring them was easy, and the UX using them was very nice.
* The worker machines running our code were (IIRC) only [2-core machines](https://docs.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#supported-runners-and-hardware-resources), and the builds were very slow.
* GA have built-in support for caching build artifacts (yay!), but the cache size is limited to 5G **per repository**. We'd needmore like 50-100G.
* Had a difficult time getting the GA Mac builds working. Not sure why. Seemed like just very slow and ended up timing out?? Not sure. Didn't debug.

I think we probably could replace Kokoro with GA, but we'd probably need to host our own GA "runners" (which we could do) and run them on bigger machines.... but doing this would nullify much of the benefit/simplicity of GA in the first place.

FWIW, other googleapis/ projects are moving lots of their builds to GA (and off of kokoro).

So, I think we experimented w/ them. At this point I don't think it makes sense to move everything to GA for our C++ builds.
